google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/ManagedCloudSdkService.java

```diff
@@ -16,35 +16,217 @@
+
+  private SdkStatus sdkStatus = SdkStatus.NOT_AVAILABLE;
+
+  private final Collection<SdkStatusUpdateListener> statusUpdateListeners = Lists.newArrayList();
+
+  private volatile ListenableFuture<Path> runningInstallationJob;
+
+  @Override
+  public void activate() {
+    // TODO track event that custom SDK is activated and used.
+
+    install();
+  }
 
   @Nullable
   @Override
   public Path getSdkHomePath() {
-    return null;
+    switch (sdkStatus) {
+      case READY:
+        return managedCloudSdk.getSdkHome();
+      default:
+        return null;
+    }
   }
 
   @Override
-  public void setSdkHomePath(String path) {}
+  public void setSdkHomePath(String path) {
+    /* unsupported, to be removed. */
+  }
 
   @Override
   public SdkStatus getStatus() {
-    return SdkStatus.NOT_AVAILABLE;
+    return sdkStatus;
   }
 
   @Override
-  public boolean installAutomatically() {
+  public boolean install() {
+    // check if installation is already running first, to prevent multiple jobs running.
+    if (runningInstallationJob == null || runningInstallationJob.isDone()) {
+      runningInstallationJob =
+          ThreadUtil.getInstance().executeInBackground(this::installSynchronously);
+      Futures.addCallback(
+          runningInstallationJob,
+          new FutureCallback<Path>() {
+            @Override
+            public void onSuccess(Path path) {
+              if (path != null) {
+                logger.info("Managed Google Cloud SDK successfully installed at: " + path);
+              }
+            }
+
+            @Override
+            public void onFailure(Throwable t) {
+              if (t instanceof InterruptedException) {
+                logger.info("Managed Google Cloud SDK installation cancelled.");
+              }
+            }
+          });
+    }
+
     return true;
   }
 
   @Override
-  public void addStatusUpdateListener(SdkStatusUpdateListener listener) {}
+  public void addStatusUpdateListener(SdkStatusUpdateListener listener) {
+    statusUpdateListeners.add(listener);
+  }
+
+  @Override
+  public void removeStatusUpdateListener(SdkStatusUpdateListener listener) {
+    statusUpdateListeners.remove(listener);
+  }
+
+  public void cancelInstall() {
+    if (runningInstallationJob != null) {
+      runningInstallationJob.cancel(true);
+    }
+  }
+
+  @VisibleForTesting
+  ManagedCloudSdk createManagedSdk() throws UnsupportedOsException {
+    return ManagedCloudSdk.newManagedSdk();
+  }
+
+  @VisibleForTesting
+  void setLogger(Logger logger) {
+    this.logger = logger;
+  }
+
+  /**
+   * Checks for Managed Cloud SDK status and creates/installs it if necessary.
+   *
+   * @return Path of the installed managed SDK, null if error occurred while installing.
+   * @throws InterruptedException if Managed Cloud SDK has been interrupted by {@link
+   *     #cancelInstall()}
+   */
+  @VisibleForTesting
+  Path installSynchronously() throws InterruptedException {
```

Comment:
this is a pretty long method. is it possible to refactor for clarity?

Reply:
Good idea. How about: `installManagedSdk()` -> calling two smaller methods `installSdk()` and `installAppEngineJava()`?

Reply:
yea. that would already be a nice improvement